include/utility/array.h

```diff
@@ -0,0 +1,197 @@
+/*************************
+ * @file array.h
+ * @author zhuzhile08 (zhuzhile08@gmail.com)
+ * 
+ * @brief a basic array implementation
+ * 
+ * @date 2022-12-24
+ * 
+ * @copyright Copyright (c) 2022
+ *************************/
+#pragma once
+
+#include <utility>
+#include <algorithm>
+
+#define NODISCARD [[nodiscard]]
+#define DEPRECATED [[deprecated]]
+
+namespace utility {
+
+// basic C-array wrapper implementation
+template <class Ty, size_t Size> struct Array {
+	typedef Ty value_type;
+	typedef value_type iterator;
+	typedef const value_type const_iterator; 
+	typedef value_type& reference;
+	typedef const value_type& const_reference;
+	typedef value_type array_type[Size];
+	typedef Array<value_type, Size> wrapper_type;
+
+	/**
+	 * @brief fill the array with the same value
+	 * 
+	 * @param value value to fill with
+	 */
+	void fill(const_reference value) { for (size_t i; i < Size; i++) m_array[i] = value; }
```

Comment:
Why are some member functions not `constexpr`?

Reply:
Didn't know void could be `constexpr`-ed, I'll add this in the rest of the API too.